hbase-server/src/main/java/org/apache/hadoop/hbase/master/compaction/CompactionOffloadManager.java

```diff
@@ -0,0 +1,134 @@
+  private CompactionOffloadSwitchStorage compactionOffloadSwitchStorage;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(CompactionOffloadManager.class.getName());
+
+  public CompactionOffloadManager(final MasterServices master) {
+    this.masterServices = master;
+    int compactionServerMsgInterval =
+        master.getConfiguration().getInt(HConstants.COMPACTION_SERVER_MSG_INTERVAL, 3 * 1000);
+    int compactionServerExpiredFactor =
+        master.getConfiguration().getInt("hbase.compaction.server.expired.factor", 2);
+    this.onlineServers = CacheBuilder.newBuilder().expireAfterWrite(
+      compactionServerMsgInterval * compactionServerExpiredFactor, TimeUnit.MILLISECONDS).build();
+    this.compactionOffloadSwitchStorage = new CompactionOffloadSwitchStorage(
+        masterServices.getZooKeeper(), masterServices.getConfiguration());
+  }
+
+  public void compactionServerReport(ServerName sn, ServerMetrics sl) {
+    this.onlineServers.put(sn, sl);
+  }
+
+  /**
+   * @return A copy of the internal list of online servers.
+   */
+  public List<ServerName> getOnlineServersList() {
+    return new ArrayList<>(this.onlineServers.asMap().keySet());
+  }
+
+  /**
+   * @return Read-only map of servers to serverinfo
+   */
+  public Map<ServerName, ServerMetrics> getOnlineServers() {
+    return Collections.unmodifiableMap(this.onlineServers.asMap());
+  }
+
+  /**
+   * May return "0.0.0" when server is not online
+   */
+  public String getVersion(ServerName serverName) {
+    ServerMetrics serverMetrics = onlineServers.asMap().get(serverName);
+    return serverMetrics != null ? serverMetrics.getVersion() : "0.0.0";
+  }
+
+  public int getInfoPort(ServerName serverName) {
+    ServerMetrics serverMetrics = onlineServers.asMap().get(serverName);
+    return serverMetrics != null ? serverMetrics.getInfoServerPort() : 0;
+  }
+
+  public IsCompactionOffloadEnabledResponse
+      isCompactionOffloadEnabled(IsCompactionOffloadEnabledRequest request) throws IOException {
+    masterServices.getMasterCoprocessorHost().preIsCompactionOffloadEnabled();
+    boolean enabled = compactionOffloadSwitchStorage.isCompactionOffloadEnabled();
+    IsCompactionOffloadEnabledResponse response = IsCompactionOffloadEnabledResponse.newBuilder()
+        .setCompactionOffloadEnabled(enabled).build();
+    masterServices.getMasterCoprocessorHost().postIsCompactionOffloadEnabled(enabled);
+    return response;
+  }
+
+  public SwitchCompactionOffloadResponse
+      switchCompactionOffload(SwitchCompactionOffloadRequest request) throws IOException {
+    boolean compactionOffloadEnabled = request.getCompactionOffloadEnabled();
+    masterServices.getMasterCoprocessorHost().preSwitchCompactionOffload(compactionOffloadEnabled);
+    boolean oldCompactionOffloadEnable =
+        compactionOffloadSwitchStorage.isCompactionOffloadEnabled();
+    if (compactionOffloadEnabled != oldCompactionOffloadEnable) {
```

Comment:
Should we add this check in the procedure implementation?

Reply:
Yes, double check to avoid duplicated procedure